### PR TITLE
Add in new HostAlloc API

### PR DESCRIPTION
**Pull request by revans2**
This fixes https://github.com/NVIDIA/spark-rapids/issues/8879

The dependency for this was merged into CUDF today, but it may be a little while before this can be merged/tested.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+  private def releasePinned(amount: Long): Unit = synchronized {
+    currentPinnedAllocated -= amount
+    if (wakeUpPinned()) {
+      wakeUpNonPinned()
+    }
+  }
+
+
+  private def releaseNonPinnedReservation(reservedAmount: Long): Unit = synchronized {
+    currentNonPinnedReserved + reservedAmount
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala:118]:
```suggestion
    currentNonPinnedReserved -= reservedAmount
```

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala [hunk 2]
```diff
@@ -0,0 +1,411 @@
...
+  }
+
+  private val pendingAllowedQueue = new HashedPriorityQueue[BlockedAllocation](100, compareBlocks)
+
+  private class BlockedAllocation(val amount: Long, val taskId: Long, val parent: AnyRef) {
+    var shouldWake = false
+    def waitUntilPossiblyReady(): Unit = {
+      shouldWake = false
+      while (!shouldWake) {
+        parent.wait(1000)
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala:49]: Are we guaranteed to be under the `parent` lock during this `wait` and the `notifyAll` later?

Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala:49]: also, is `parent` normally `HostAlloc`? In which cases would `parent` be something else?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+
+
+  private def tryReservePinned(amount: Long): Option[HostMemoryReservation] = {
+    val ret = Option(PinnedMemoryPool.tryReserve(amount))
+    ret.map { reservation =>
+      new WrappedPinnedReservation(reservation)
+    }
+  }
+
+  private object UnlimitedReservation extends HostMemoryReservation {
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala:153]: in which cases would we use an unlimited reservation?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala (hunk 2 above)
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala:49]: Yes it is required or else there will be an error from the JVM.  parent right now is always `HostAlloc`. We could try and switch it over to something more like `HostAlloc.this.wait(1000)`

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    override def allocate(amount: Long): HostMemoryBuffer = synchronized {
+      if (amount > reservedAmount) {
+        throw new OutOfMemoryError("Could not allocate. Remaining memory reservation is " +
+            s"too small $amount out of $reservedAmount")
+      }
+      val buf = tryAllocNonPinned(amount, fromReservation = true)
+      buf.foreach { b =>
+        reservedAmount -= align(b.getLength)
+      }
+      buf.get
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala:179]: It should only really fail if an internal error happened. I am probably trying to reuse too much from tryAllocNonPinned without refactoring/renaming it.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+package com.nvidia.spark.rapids
+
+import java.util.Comparator
+
+import ai.rapids.cudf.{ColumnView, HostMemoryBuffer, HostMemoryReservation, MemoryBuffer, PinnedMemoryPool}
+import com.nvidia.spark.rapids.HostAlloc.align
+
+import org.apache.spark.TaskContext
+
+private class HostAlloc(nonPinnedLimit: Long) {
```
[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala:26]: If you really want me to I will. Is it unclear the type of memory being allocated?

[the remark this answers is not in the document]
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala:26]: If you really want me to I will. Is it unclear what this does?

**Comment by revans2** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala [hunk 6]
```diff
@@ -64,6 +82,10 @@ private class HostAlloc(nonPinnedLimit: Long) {
     }
   }
 
+  /**
+   * A callback so we know when a pinned host buffer was released.
+   * @param amount
```
Comment (jlowe) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala:87]: This should either have a description or be removed.

**Comment by revans2** [comment]
build

**Review by jlowe — APPROVED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala:26]: the main nit I had was that HostAlloc sounded more like a product of an allocator than the allocator itself, but it's just a nit. Not a must fix.

[the remark this answers is not in the document]
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala:26]: It's a nit, not a must fix. It is clear to me that non-pinned is pageable.

**Review by abellina — APPROVED** [review]

**Pull request merged**